server/src/main/java/org/elasticsearch/plugins/PluginsLoader.java

```diff
@@ -113,30 +113,41 @@ public static LayerAndLoader ofLoader(ClassLoader loader) {
     private final Map<String, LoadedPluginLayer> loadedPluginLayers;
 
     /**
-     * Constructs a new PluginsLoader
+     * Constructs a new PluginsLoader. With this factory method, the set of qualified exports/opens to Java modular modules/plugins will
+     * include all qualified exports from the server module. This is the desired behaviour in production use.
      *
      * @param modulesDirectory The directory modules exist in, or null if modules should not be loaded from the filesystem
      * @param pluginsDirectory The directory plugins exist in, or null if plugins should not be loaded from the filesystem
      */
     public static PluginsLoader createPluginsLoader(Path modulesDirectory, Path pluginsDirectory) {
-        return createPluginsLoader(modulesDirectory, pluginsDirectory, true);
+        Map<String, List<ModuleQualifiedExportsService>> qualifiedExports = new HashMap<>(ModuleQualifiedExportsService.getBootServices());
+        addServerExportsService(qualifiedExports);
+        return createPluginsLoader(modulesDirectory, pluginsDirectory, qualifiedExports);
+    }
+
+    /**
+     * Constructs a new PluginsLoader. With this factory method, the set of qualified exports/opens to Java modular modules/plugins will
+     * not include qualified exports from the server module. This is the desired behaviour in test use, as tests are running non-modular.
+     *
+     * @param modulesDirectory The directory modules exist in, or null if modules should not be loaded from the filesystem
+     * @param pluginsDirectory The directory plugins exist in, or null if plugins should not be loaded from the filesystem
+     */
+    public static PluginsLoader createPluginsLoaderWithoutServerExports(Path modulesDirectory, Path pluginsDirectory) {
```

Comment:
Why do we need 3 different public variants of creating a PluginsLoader? Shouldn't there be just 1?